base/bucket_n1ql.go

```diff
@@ -23,38 +33,89 @@ type IndexMetadata struct {
+	waitTime := 10 * time.Millisecond
+	for i := 1; i <= MaxQueryRetries; i++ {
+		queryResults, queryErr := bucket.ExecuteN1qlQuery(n1qlQuery, params)
+		if queryErr == nil {
+			return queryResults, queryErr
+		}
+
+		// Timeout error - return named error
+		if isGoCBTimeoutError(queryErr) {
+			return queryResults, ErrViewTimeoutError
+		}
 
-	if isGoCBTimeoutError(err) {
-		return results, ErrViewTimeoutError
+		// Non-retry error - return
+		if !isIndexerError(queryErr) {
+			Warn("Error when querying index using statement: [%s]", bucketStatement)
+			return queryResults, queryErr
+		}
+
+		// Indexer error - wait then retry
+		err = queryErr
+		Warn("Indexer error during query - retry %d/%d", i, MaxQueryRetries)
+		time.Sleep(waitTime)
+		waitTime = time.Duration(waitTime * 2)
 	}
 
-	return results, err
+	Warn("Exceeded max retries for query when querying index using statement: [%s], err:%v", bucketStatement, err)
+	return nil, err
 }
 
 // CreateIndex creates the specified index in the current bucket using on the specified index expression.
-func (bucket *CouchbaseBucketGoCB) CreateIndex(indexName string, expression string, numReplica uint) error {
+func (bucket *CouchbaseBucketGoCB) CreateIndex(indexName string, expression string, filterExpression string, options *N1qlIndexOptions) error {
 
 	createStatement := fmt.Sprintf("CREATE INDEX %s ON %s(%s)", indexName, bucket.GetName(), expression)
 
-	if numReplica > 0 {
-		createStatement = fmt.Sprintf("%s with {num_replica:%d}", createStatement, numReplica)
+	// Add filter expression, when present
+	if filterExpression != "" {
+		createStatement = fmt.Sprintf("%s WHERE %s", createStatement, filterExpression)
+	}
+
+	// Replace any BucketQueryToken references in the index expression
+	createStatement = strings.Replace(createStatement, BucketQueryToken, bucket.GetName(), -1)
+
+	return bucket.createIndex(createStatement, options)
+}
+
+// CreateIndex creates the specified index in the current bucket using on the specified index expression.
+func (bucket *CouchbaseBucketGoCB) CreatePrimaryIndex(indexName string, options *N1qlIndexOptions) error {
+
+	createStatement := fmt.Sprintf("CREATE PRIMARY INDEX %s ON %s", indexName, bucket.GetName())
+	return bucket.createIndex(createStatement, options)
+}
+
+func (bucket *CouchbaseBucketGoCB) createIndex(createStatement string, options *N1qlIndexOptions) error {
+
+	if options != nil {
+		withClause, marshalErr := json.Marshal(options)
+		if marshalErr != nil {
+			return marshalErr
+		}
+		createStatement = fmt.Sprintf(`%s with %s`, createStatement, withClause)
 	}
 
 	LogTo("Index+", "Attempting to create index using statement: [%s]", createStatement)
 	n1qlQuery := gocb.NewN1qlQuery(createStatement)
 	results, err := bucket.ExecuteN1qlQuery(n1qlQuery, nil)
-	if err != nil {
+	if err != nil && !IsRecoverableCreateIndexError(err) {
+		Warn("Error creating index with statement: %s, error:%v", createStatement, err)
```

Comment:
This seems like a good place to wrap the error in order to get the stack trace and also let the handler of the error decide whether to print anything to the logs or not.